### ISSUE TITLE
feedback from tiagoantao 

### ISSUE THREAD

**Issue by martenson** [galaxyproject/planemo#382]
from @tiagoantao on using Planemo

I had several problems. I still have to be quite frank. Note that some problems below are not documentation related:

Understanding the correct workflow. When should I use test, when should I use shed_test?
What are the parameters for testing? The documentation (which is just a pull of the --help output) is not very clear about that. Sometimes I did not know if I was testing all the tools in a directory if or a single tool (this was particularly bad when I started working on a fork of tools-iuc).

The speed of shed_test (with all the overheads)

The output of the tests. For example, some cases the XML files were not correct, no help there. Also the format of the output on HTML is not helpful.

Suggestions:
Put some concrete examples of usage in the documentation. Examples with shed_test and test.
Make suggestions on planemo workflows.
Run planemo lint before testing (automatically). I spent hours testing things that I should have linted. While this was a mistake of mine, the tool could actually have made life easier here.
Easier test output to interpret...

**Comment by peterjc** [comment]
There are a lot of good points here - not least of which is running the functional tests takes too long. There an old Trello issue https://trello.com/c/wL21d2do/1017-functional-tests-take-too-long whose scope was narrowed to speeding up the test database creation and schema migration - addressed by @jmchilton with things like:

```
export GALAXY_TEST_DB_TEMPLATE=https://github.com/jmchilton/galaxy-downloads/raw/master/db_gx_rev_0127.sqlite
```

On a more positive note, with my TravisCI setup the recent dev branch does seem to be consistently faster than the current master (and by proxy, the current stable releases) at running the tests, e.g.

* https://travis-ci.org/peterjc/galaxy_blast/builds/92018003
* https://travis-ci.org/peterjc/pico_galaxy/builds/91567501
* https://travis-ci.org/peterjc/galaxy_mira/builds/89677151

As I mentioned to @natefoo via Twitter, part of this is the improved Python dependency setup time (eggs vs wheels), but that is only a small part of the above speed up.

**Comment by martenson** [comment]
I like the point of tying lint & test, i.e. not running the tests if linting errors.

**Comment by peterjc** [comment]
Yes, including a pre-lint automatically prior to a test makes sense (as long as it is very quick).

**Comment by peterjc** [comment]
Re: *The output of the tests. For example, some cases the XML files were not correct, no help there. Also the format of the output on HTML is not helpful.*

That could possibly be related to https://github.com/galaxyproject/planemo/issues/306 or https://github.com/galaxyproject/planemo/issues/381 ?

When it works, the HTML output from planemo is far nicer than trying to spot the failures at the terminal from Galaxy's simple ``run_tests.sh`` - which is not to say there isn't room for improvement. I'm curious what @tiagoantao had in mind here.

**Comment by jmchilton** [comment]
So the problem I don't think was introduce in Planemo 0.19.0 to 0.20.0 - I think it was that Galaxy changed from eggs to wheels and virtualenv because a hard dependency for Galaxy.

In the early days of planemo it made a lot of sense for it to target the dev branch of Galaxy - the people mostly using planemo were people who also frequently run dev and are used to these problems (and in fact want to catch them before Galaxy releases.) It might be time to switch planemo to target master by default and make targeting dev a flag. I'll create an issue for this.

Continuing to be defensive :expressionless:... ``shed_test``` does stink pretty badly - but because of tool shed problems, not planemo ones - it is much better (easier, faster, prettier, robust) than what it replaced (manually running the install and test framework). There is a lot of tension here with testing and everyone who has tried shed_test has encountered this I think - neither ``test`` nor ``shed_test`` is really what people want to do - but what people want to do requires huge changes to the tool shed that I'm not sure anyone is in the position to make those changes. I have put a lot of effort into shed_test documentation trying to explain what it is trying to do and why it is different test, but I might be too close to the differences - someone else should take a crack at revising that documentation. There should also a high-level description of the differences somewhere - possible in the shed workflow documentation (http://planemo.readthedocs.org/en/latest/writing_advanced.html#publishing-to-the-tool-shed).

**Comment by nsoranzo** [comment]
@martenson wrote:
> I like the point of tying lint & test, i.e. not running the tests if linting errors.

Just to be clear: tests should not be run if `planemo lint --fail_level error` fails.

**Comment by jmchilton** [comment]
There are tools that will fail linting and will run fine: e.g.

```
    if "version" not in root.attrib:
        lint_ctx.error("Tool does not define a version attribute.")
```

But I am still open to do this - it should be enabled by default right? Are these the three options we want?

- ```--skip_lint```: By default planemo will lint tools for errors before testing, disable this check with this flag.
- ```--lint_fail_level```: Skip testing if lint fails at this level (ERROR, WARN), ignored if --skip_lint is specified.
- ```--lint_skips```: Comma separate list of linters to skip, ignored if --skip_lint is specified.

**Comment by erasche** [comment]
Is this really necessary? It feels like conflating two separate functions that end users should be keeping in a makefile.

```Makefile
lint:
    planemo lint path/to/data

test: lint
    planemo test ...
```

**Comment by peterjc** [comment]
That's kind of what I do with my ``.travis.yml`` setup. Is there a simple output file from ``planemo lint`` which can be used with make rules to re-run automatically when needed?

It would fit well under *Make suggestions on planemo workflows*, document lint first as how planemo is intended to be used.

**Comment by jmchilton** [comment]
You know what might be better than adding linting to test by default - #17.

**Comment by erasche** [comment]
:+1:

**Comment by tiagoantao** [comment]
Another (minor) annoyance that I have when moving things to tools-iuc is the repository information, for example:
Imagine that you have package_x, a new one. That package will be under the user (eg tiagoantao), but when ported to tools-iuc it will have to be under iuc. The dependency will be going backwards and forwards during the testing review phase.
This is currently happening withe package_raxml_8_2_4 while I incorporate feedback from @bgruening .

<tool_dependency>
  <package name="raxml" version="8.2.4">
    <repository name="package_raxml_8_2_4" owner="tiagoantao" />


All in all there are a plethora of minor issues that when compounded make the planemo experience sub-standard.

But, I think my biggest complaint is easy to solve: document a best-practice workflow when testing tools. I am still quite confused...
